src/outputs/harbor_output.ml

```diff
@@ -204,59 +202,78 @@ module Make (T : T) = struct
-        c.metapos <- 0 ;
-        process "\000" rem after )
-      else (
-        c.metapos <- String.length after ;
-        Printf.sprintf "%s%s%s%s" rem before meta after )
+      let before = Strings.sub data 0 pos in
+      let after = Strings.sub data pos (Strings.length data - pos) in
+      if Strings.length after > c.metaint then
+        (
+          let rem = Strings.concat [rem;before;Strings.of_string meta] in
+          c.metapos <- 0 ;
+          process "\000" rem after
+        )
+      else
+        (
+          c.metapos <- Strings.length after;
+          Strings.concat [rem;before;Strings.of_string meta;after]
+        )
     in
     if c.metaint > 0 then
-      if String.length data + c.metapos > c.metaint then
-        let meta =
-          Tutils.mutexify c.meta.metadata_m (fun () -> c.meta.metadata) ()
-        in
-        process (get_meta meta) "" data
-      else (
-        c.metapos <- c.metapos + String.length data ;
-        data )
+      let datalen = Strings.length data in
+      if datalen + c.metapos > c.metaint then
+        let meta = Tutils.mutexify c.meta.metadata_m (fun () -> c.meta.metadata) () in
+        process (get_meta meta) Strings.empty data
+      else
+        (
+          c.metapos <- c.metapos + datalen;
+          data
+        )
     else data
 
   let rec client_task c =
     let __pa_duppy_0 =
       Duppy.Monad.Io.exec ~priority:Tutils.Maybe_blocking c.handler
         (Tutils.mutexify c.mutex
            (fun () ->
-             let buflen = Buffer.length c.buffer in
+             let buflen = Strings.length c.buffer in
              let data =
-               if buflen > c.chunk then (
-                 let data = Some (add_meta c (Buffer.contents c.buffer)) in
-                 Buffer.reset c.buffer ; data )
-               else None
+               if buflen > c.chunk then
+                 let data = add_meta c c.buffer in
+                 c.buffer <- Strings.empty;
+                 data
+               else Strings.empty
              in
              Duppy.Monad.return data )
            ())
     in
     Duppy.Monad.bind __pa_duppy_0 (fun data ->
         Duppy.Monad.bind
-          ( match data with
-          | None ->
+          (
+            if Strings.is_empty data then
               Duppy.Monad.bind (Duppy_m.lock c.condition_m) (fun () ->
                   Duppy.Monad.bind (Duppy_c.wait c.condition c.condition_m)
                     (fun () -> Duppy_m.unlock c.condition_m ) )
-          | Some s ->
-              Duppy.Monad.Io.write ?timeout:(Some c.timeout)
-                ~priority:Tutils.Non_blocking c.handler (Bytes.of_string s) )
+            else
+              let rec write = function
+                | s::l ->
+                  Duppy.Monad.bind
+                    (Duppy.Monad.Io.write ?timeout:(Some c.timeout)
+                       ~priority:Tutils.Non_blocking c.handler
+                       (* TODO: can we use unsafe_of_string here? *)
```

Comment:
@toots I would expect `Duppy.Monad.Io.write` to take a `string` not a `bytes`

Reply:
I think that this is mimicking the `Unix` API, which also takes a `bytes` argument now.

Reply:
And, yes, `unsafe_of_string` is totally acceptable I believe.